provisioning/transport/http/src/http.ts

```diff
@@ -0,0 +1,251 @@
+import { RestApiClient, Http as Base } from 'azure-iot-http-base';
+import { errors, X509 } from 'azure-iot-common';
+import * as machina from 'machina';
+
+export type ProvisioningResponseCallback = (err: Error, result?: any) => void;
+
+const _defaultHeaders = {
+  'Accept' : 'application/json',
+  'Content-Type' : 'application/json; charset=utf-8'
+};
+
+export class Http extends EventEmitter {
+
+  private _idScope: string;
+  private _registrationId: string;
+  private _body: any;
+  private _forceRegistration: boolean;
+  private _operationId: string;
+
+  private _fsm: machina.Fsm;
+  private _apiVersion: string = '2017-08-31-preview';
+  private _provisioningServer: string = 'global.azure-devices-provisioning.net';
+  private _httpTimeout: number = 4000;
+  private _operationStatusPollingInterval: number = 2000;
+  private _pollingTimer: any;
+  private _restApiClient: RestApiClient;
+  private _registrationCallback: ProvisioningResponseCallback;
+
+  /**
+   * @private
+   * @constructor
+   * @param config The configuration object.
+   */
+  /* Codes_SRS_NODE_PROVISIONING_HTTP_18_001: [ The `Http` constructor shall accept the following properties:
+  - `idScope` - a string specifiying the scope of the provisioning operations,
+  - `registrationId` - the registration id for the specific device ] */
+  constructor(idScope: string, registrationId: string, httpBase?: Base) {
+    super();
+    this._idScope = idScope;
+    this._registrationId = registrationId;
+
+    this._fsm = new machina.Fsm({
+      namespace: 'http-provisioning',
+      initialState: 'idle',
+      states: {
+        idle: {
+          _onEnter: (callback, err, response) => {
+            this._pollingTimer = null;
+            this._restApiClient = null;
+            this._body = null;
+            this._forceRegistration = false;
+            this._operationId = '';
+            this._registrationCallback = null;
+            if (callback) {
+              callback(err, response);
+            }
+          },
+          register: (authorization, forceRegistration, body, callback) => {
+            this._registrationCallback = callback;
+            this._forceRegistration = forceRegistration;
+            this._body = body;
+            if (authorization.hasOwnProperty('cert')) {
+              this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'x509' : authorization}, 'azure-node-provisioning', httpBase);
+            } else {
+              this._restApiClient = new RestApiClient({ 'host' : this._provisioningServer , 'sharedAccessSignature' : authorization}, 'azure-node-provisioning', httpBase);
+            }
+            this._fsm.transition('registering', callback);
+          },
+          disconnect : (callback) => {
+            // nothing to do.
+            callback();
+          }
+        },
+        registering: {
+          _onEnter: (callback) => {
+            this._sendRegistrationRequest((err, responseBody) => {
```

Comment:
>responseBody [](start = 48, length = 12)

not sure due to naming: is it the object, or the JSON string that still needs to be parsed?